### DISCUSSION TITLE
Feedback request for Miller 6 docs

### DISCUSSION THREAD

**Discussion by johnkerl** [johnkerl/miller#621] [Q&A]
Within the context of the Go port ([README](https://github.com/johnkerl/miller/blob/main/go/README.md), [tracking issue](https://github.com/johnkerl/miller/issues/372), and  responses from the 2021 Miller survey) I'd like to hear some thoughts on Miller 6 docs.

* URL is https://johnkerl.org/miller6. (Ultimately these will replace https://miller.readthedocs.io/en/latest, once Miller 6 is released but not yet)
* Styling is (I believe) much better for desktop as well as mobile -- in terms of access to menu/navbar
* There's more, and gentler, up-front material for non-experts; expert-level material is pushed farther down
* Many big pages have been split into smaller ones -- including the non-DSL and DSL reference pages, those two are now many. (Maybe _too_ many -- ?)
* Also note not all content has been updated for Miller 6 -- there are statements in some pages which are true for Miller 5, not true for Miller 6 anymore. Also several spots I've marked with "TODO". In short this is a good draft but not a final one by any means.

Anyway, those are _my_ perspectives -- what about yours?

**Comment by aborruso** [comment]
Hi @johnkerl this is great, thank you!!

Some points:

- MkDocs is a really great tool. I'm using it for my Italian Miller guide (still too short to show you);
  - there is a really great theme which I like a lot more than RTD one. It's Material https://squidfunk.github.io/mkdocs-material/
  - I have discovered this theme studying csvtk documentation https://bioinf.shenwei.me/csvtk/
- a great thing, using something like MkDocs, is to collaborate on documentation because it is very very easy;
- I love that you have inserted more, and gentler, up-front material for non-experts. I am that kind of user, and I think it's important that a gem like Miller also contains this type of content;
- if you don't change the style, there are things like this to fix

![image](https://user-images.githubusercontent.com/30607/128481855-b60589cf-d66b-4b25-b7e1-0ab81139d78f.png)


Thank you very much

**Reply by johnkerl** [reply]
Thanks @aborruso !!

Looking forward to your Italian docs, would love to see a peek!

I'll definitely check out 'material', thank you! :) 

Regarding line-wraps for pre-formatted text -- yes, there are a lot of those ... at present I think they'll take a round of manual edits. It's possible to fix these using CSS with the `white-space: pre-wrap;` but I'm undecided whether to use that. The plus is that that CSS tweak would automatically wrap long lines without any manual editing. The minus is I fear people would try to copy/paste command-lines into their terminal & they wouldn't work anymore. So maybe I should manually insert line breaks with ` \` at the end ... .

What do you think?

**Reply by johnkerl** [reply]
@aborruso I experimented a bit and the CSS-wrap method will work fine :)

When I select a line-wrapped bit in the browser and copy, then paste into the terminal, it pastes as a single line (as desired):

<img width="636" alt="Screen Shot 2021-08-06 at 10 15 30 AM" src="https://user-images.githubusercontent.com/2008794/128524074-79e800ea-5177-4cbe-bf35-29b7f618f2cb.png">

<img width="642" alt="Screen Shot 2021-08-06 at 10 18 13 AM" src="https://user-images.githubusercontent.com/2008794/128524236-3877e664-cd03-41e1-b08f-ce0ef2f4a53a.png">

**Reply by johnkerl** [reply]
Also @aborruso thanks for the tip on the `material` theme -- it also does great on desktop+mobile, but handles nested TOC-tree better than the `readthedocs` theme does.

After some CSS fixups (fonts, colors) I'll push this as the theme for the Miller6 docs. :D

**Reply by johnkerl** [reply]
@aborruso https://github.com/johnkerl/miller/pull/624 switches to `material` theme. I like this much better -- thanks again! :)

**Reply by johnkerl** [reply]
P.S. you can reload https://johnkerl.org/miller6 to pick up the changes

**Reply by aborruso** [reply]
Dear @johnkerl I'm very very proud of this. I love this project and I try to give my contribution. I'm sorry I can't do great things.

Material is really great:

- try in example to open from desktop this new Miller documentation and then click on <kbd>N</kbd> and then <kbd>P</kbd>;
- I love 
  - [abbrevations](https://squidfunk.github.io/mkdocs-material/reference/abbreviations/#abbreviations_1);
  - [admonitions](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#usage);
  - [content tabs](https://squidfunk.github.io/mkdocs-material/reference/content-tabs/#usage);
  - [footnotes](https://squidfunk.github.io/mkdocs-material/reference/footnotes/#usage);
  - [meta tags](https://squidfunk.github.io/mkdocs-material/reference/meta-tags/#usage).

I have some points about this new Material version. Can I add it here in the next days?

The CSS-wrap is OK, thank you.

Have a nice week-end

**Reply by johnkerl** [reply]
> I have some points about this new Material version. Can I add it here in the next days?

Yes, please! :D

**Reply by johnkerl** [reply]
> I love this project and I try to give my contribution. I'm sorry I can't do great things.

@aborruso you have done countless great things! :D

**Reply by trantor** [reply]
As a feedback, I do like the theme quite a bit.
On another note, I've seen a lot of activity over the last few weeks and I've been quite annoyed with myself for not having being able to lend a hand in any way given my daily activities.
Do you need some help with proofreading the docs (I am very good at spotting inconsistencies) or some other kind of feedback on code @johnkerl , by any chance?

**Reply by johnkerl** [reply]
@trantor yes I would love that!!!!!! Whatever you have time for. :)

One caveat ... I'm midway through working through https://github.com/johnkerl/miller/blob/main/docs6/docs/proofreads.txt ... so if you don't mind looking and finding things that I may already be editing, go right ahead! Or if you'd rather be more careful about duplication of effort, maybe hold off a couple weeks -- totally up to you.

Code -- similarly; there are many things I feel 'ready for review' in Miller 6 but lots of things marked `TODO`. Any feedback you have would be great. The caveat for `docs6` is that known to-do items are in the single `docs6/docs/proofreads.txt`; for the Go code, the to-do items are distributed throughout the source code with either `TODO` or `xxx`.

**Reply by NikosAlexandris** [reply]
> @aborruso I experimented a bit and the CSS-wrap method will work fine :)
> 
> When I select a line-wrapped bit in the browser and copy, then paste into the terminal, it pastes as a single line (as desired):
> 
> <img alt="Screen Shot 2021-08-06 at 10 15 30 AM" width="636" src="https://user-images.githubusercontent.com/2008794/128524074-79e800ea-5177-4cbe-bf35-29b7f618f2cb.png">
> 
> <img alt="Screen Shot 2021-08-06 at 10 18 13 AM" width="642" src="https://user-images.githubusercontent.com/2008794/128524236-3877e664-cd03-41e1-b08f-ce0ef2f4a53a.png">

Byt the way, the colorised output looks awesome -- really useful and easier to read on-screen, given the dark terminal background.-

**Comment by NikosAlexandris** [comment]
Looks awesome! May I propose a Glossary page too? A reference to have acronyms fully spelled out and descriptive sentences. A tutorial and examples are really the Thing. Yet a glossary is a quick overview to grasp the terminology. And there are quite some _keywords_ there to list.

**Reply by NikosAlexandris** [reply]
Really @johnkerl the new docs is just fantastic -- makes reading so much more enjoyable.

**Reply by johnkerl** [reply]
Awesome @NikosAlexandris thank you!! :D

**Reply by johnkerl** [reply]
@NikosAlexandris https://johnkerl.org/miller6/glossary/index.html is a start ... please let me know what you think of the general direction.

**Reply by NikosAlexandris** [reply]
I needed this, quite many times! :-)

Just my personal preference of learning/understanding, is to really have an acronym fully spelled out in the glossary. Like, it took me days to finally understand Domain-Specific-Language. And, maybe even still today I am confused about the D in the DKVP. Though, it's clear to me now, and at least, that KVP stands for key-value-pair :-)

**Comment by johnkerl** [comment]
Sorry @NikosAlexandris -- DKVP is for 'delimited key-value pairs' -- fixed in the glossary -- thank you! :)

**Comment by johnkerl** [comment]
Also @NikosAlexandris was it you that asked about multicore -- ?

https://johnkerl.org/miller6/cpu/

**Reply by NikosAlexandris** [reply]
Maybe indirectly in https://github.com/johnkerl/miller/discussions/637 (?) -- Thank you Anyway :-)

**Reply by johnkerl** [reply]
Found it, it was a response in the [survey](https://docs.google.com/forms/d/e/1FAIpQLSfNgLS9WVRq9mu8dZlMbS7LbTyRH1diRDnT_dGiavSOh_A8xA/viewform?usp=sf_link) :)

**Comment by NikosAlexandris** [comment]
Additional Markdown tables for the **data-conversion keystroke-savers**: https://gist.github.com/NikosAlexandris/65a88b4df50c33f16ff2dba16c8ef474

**Reply by johnkerl** [reply]
@NikosAlexandris https://johnkerl.org/miller6/reference-main-flag-list/index.html#format-conversion-keystroke-saver-flags 💥

**Reply by aborruso** [reply]
> @NikosAlexandris https://johnkerl.org/miller6/reference-main-flag-list/index.html#format-conversion-keystroke-saver-flags 💥

Fantastic!!!

**Comment by NikosAlexandris** [comment]
The [Data-conversion keystroke-savers](https://johnkerl.org/miller6/file-formats/#data-conversion-keystroke-savers) section seems currently to _not_ use a new/Go version of miller, so `mlr help format-conversion` fails.

**Reply by johnkerl** [reply]
@NikosAlexandris the docs are produced by code on a branch which I haven't merged to head yet ... sorry for the confusion. :^/

**Reply by NikosAlexandris** [reply]
If such comments cause you to waste more time, rather than save and be useful, let me please know -- I do know it's stuff under construction. I will be more conservative then to report such minor issues.

**Reply by johnkerl** [reply]
@NikosAlexandris not at all!! This is all super-helpful, more please! :) I let you about the branch because I felt bad about having confused you . :)

**Reply by NikosAlexandris** [reply]
Great. I can create mini Pull-Requests also, hopefully saving some of your precious time.

**Comment by NikosAlexandris** [comment]
@johnkerl I read through https://johnkerl.org/miller6/programming-language/#out-of-stream-variables-begin-and-end and I was quite pleased with the new style, and (obviously after having spent some time with Miller by now) started to digest things faster. One thing I miss reading, to know it can work or even how it will work, is using `put` with some existing field names to create a/some new field/s, then use the newly created fields in a subsequent statement/s. Like, what I am trying to do in https://github.com/johnkerl/miller/discussions/651#discussioncomment-1292379 but just a simple case.

The _intro_ tells us, and shows, how to write a simple statement, that multiple statements are to be separated by `;` and that multi-line statements are possible too. Then, except for the out-of-stream variables (which indeed and naturally appear twice or more (?) in a statement), there is no single simple example that demonstrates that a new field can be reused further in a subsequent statement. I think such a simple example deserves to be presented in the _intro_.

Example:
```bash
➜  echo 'a=3' |mlr.go --d2c put '$b = $a - 1 ; $c = $a - $b'
a,b,c
3,2,1
```

Sorry if I missed to find it or please ignore if you see not fit for it in the _intro_.

---

ps- Off-topic on Miller syntax + GitHub: would be cool to be able to apply Miller's syntax highlighting in issues and elsewhere in this repository via:
```` ```miller````
`..`
```` ``` ````

**Reply by johnkerl** [reply]
@NikosAlexandris how is this?

https://johnkerl.org/miller6/10min/#computing-new-fields

<img width="1240" alt="Screen Shot 2021-09-08 at 12 00 10 AM" src="https://user-images.githubusercontent.com/2008794/132444400-941c73c8-f4df-4fe8-b8e9-829b0e3783c0.png">

**Reply by NikosAlexandris** [reply]
Feels natural to have it right after the 1st `put` example.
